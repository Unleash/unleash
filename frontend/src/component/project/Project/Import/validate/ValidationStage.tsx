import { ImportLayoutContainer } from '../ImportLayoutContainer';
import { Box, Button, styled, Typography } from '@mui/material';
import React, { FC, useEffect, useState } from 'react';
import {
    IValidationSchema,
    useValidateImportApi,
} from 'hooks/api/actions/useValidateImportApi/useValidateImportApi';
import { ConditionallyRender } from 'component/common/ConditionallyRender/ConditionallyRender';
import useToast from 'hooks/useToast';
import { formatUnknownError } from 'utils/formatUnknownError';
import { ActionsContainer } from '../ActionsContainer';
import { IMPORT_CONFIGURATION_BUTTON } from 'utils/testIds';
import PermissionButton from 'component/common/PermissionButton/PermissionButton';
import { CREATE_FEATURE } from 'component/providers/AccessProvider/permissions';
import { usePlausibleTracker } from 'hooks/usePlausibleTracker';

const ImportInfoContainer = styled(Box)(({ theme }) => ({
    backgroundColor: theme.palette.background.elevation2,
    borderRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(3),
}));

const Label = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.secondary,
}));
const Value = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.primary,
    fontWeight: theme.fontWeight.bold,
}));

const ErrorContainer = styled(Box)(({ theme }) => ({
    border: `1px solid ${theme.palette.error.border}`,
    borderRadius: theme.shape.borderRadiusLarge,
    paddingBottom: theme.spacing(2),
}));

const WarningContainer = styled(Box)(({ theme }) => ({
    border: `1px solid ${theme.palette.warning.border}`,
    borderRadius: theme.shape.borderRadiusLarge,
    paddingBottom: theme.spacing(2),
}));

const ErrorHeader = styled(Box)(({ theme }) => ({
    color: theme.palette.error.dark,
    backgroundColor: theme.palette.error.light,
    fontSize: theme.fontSizes.smallBody,
    borderBottom: `1px solid ${theme.palette.error.border}`,
    borderTopLeftRadius: theme.shape.borderRadiusLarge,
    borderTopRightRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(2),
}));

const WarningHeader = styled(Box)(({ theme }) => ({
    color: theme.palette.warning.dark,
    backgroundColor: theme.palette.warning.light,
    fontSize: theme.fontSizes.smallBody,
    borderBottom: `1px solid ${theme.palette.warning.border}`,
    borderTopLeftRadius: theme.shape.borderRadiusLarge,
    borderTopRightRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(2),
}));

const ErrorMessage = styled(Box)(({ theme }) => ({
    color: theme.palette.error.dark,
    fontSize: theme.fontSizes.smallBody,
}));

const WarningMessage = styled(Box)(({ theme }) => ({
    color: theme.palette.warning.dark,
    fontSize: theme.fontSizes.smallBody,
}));

const StyledItems = styled('ul')(({ theme }) => ({
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(0),
    paddingLeft: theme.spacing(3),
    paddingBottom: theme.spacing(3),
    borderBottom: `1px dashed ${theme.palette.neutral.border}`,
}));

const StyledItem = styled('li')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
}));

export const ValidationStage: FC<{
    environment: string;
    project: string;
    payload: string;
    onClose: () => void;
    onSubmit: () => void;
    onBack: () => void;
}> = ({ environment, project, payload, onClose, onBack, onSubmit }) => {
    const { validateImport } = useValidateImportApi();
    const { setToastData } = useToast();
    const { trackEvent } = usePlausibleTracker();
    const [validationResult, setValidationResult] = useState<IValidationSchema>(
        { errors: [], warnings: [], permissions: [] }
    );
    const [validJSON, setValidJSON] = useState(true);

    const trackValidation = (result: IValidationSchema) => {
        if (result.errors.length > 0 || result.permissions.length > 0) {
            trackEvent('export_import', {
                props: {
                    eventType: `validation fail`,
                },
            });
        } else {
            trackEvent('export_import', {
                props: {
                    eventType: `validation success`,
                },
            });
        }
        setValidationResult(result);
    };

    useEffect(() => {
        validateImport({ environment, project, data: JSON.parse(payload) })
            .then(trackValidation)
            .catch(error => {
                setValidJSON(false);
                setToastData({
                    type: 'error',
                    title: formatUnknownError(error),
                });
            });
    }, []);

    return (
        <ImportLayoutContainer>
            <ImportInfoContainer>
                <Typography sx={{ mb: 1.5 }}>
                    You are importing this configuration in:
                </Typography>
                <Box sx={{ display: 'flex', gap: 3 }}>
                    <span>
                        <Label>Environment: </Label>
                        <Value>{environment}</Value>
                    </span>
                    <span>
                        <Label>Project: </Label>
                        <Value>{project}</Value>
                    </span>
                </Box>
            </ImportInfoContainer>
            <ConditionallyRender
                condition={validationResult.permissions.length > 0}
                show={
                    <ErrorContainer>
                        <ErrorHeader>
                            <strong>Missing permissions!</strong> There are some
                            permissions that you need to be granted before
                            importing this configuration
                        </ErrorHeader>
                        {validationResult.permissions.map(error => (
                            <Box key={error.message} sx={{ p: 2 }}>
                                <ErrorMessage>{error.message}</ErrorMessage>
                                <StyledItems>
                                    {error.affectedItems.map(item => (
                                        <StyledItem key={item}>
                                            {item}
                                        </StyledItem>
                                    ))}
                                </StyledItems>
                            </Box>
                        ))}
                    </ErrorContainer>
                }
            />
            <ConditionallyRender
                condition={validationResult.errors.length > 0}
                show={
                    <ErrorContainer>
                        <ErrorHeader>
                            <strong>Conflict!</strong> There are some errors
                            that need to be fixed before the import.
                        </ErrorHeader>
                        {validationResult.errors.map(error => (
                            <Box key={error.message} sx={{ p: 2 }}>
                                <ErrorMessage>{error.message}</ErrorMessage>
                                <StyledItems>
                                    {error.affectedItems.map(item => (
                                        <StyledItem key={item}>
                                            {item}
                                        </StyledItem>
                                    ))}
                                </StyledItems>
                            </Box>
                        ))}
                    </ErrorContainer>
                }
            />
            <ConditionallyRender
                condition={validationResult.warnings.length > 0}
                show={
                    <WarningContainer>
                        <WarningHeader>
                            <strong>Warning!</strong> It is recommended to
                            verify the following information before importing.
                        </WarningHeader>
                        {validationResult.warnings.map(warning => (
                            <Box key={warning.message} sx={{ p: 2 }}>
                                <WarningMessage>
                                    {warning.message}
                                </WarningMessage>
                                <StyledItems>
                                    {warning.affectedItems.map(item => (
                                        <StyledItem key={item}>
                                            {item}
                                        </StyledItem>
                                    ))}
                                </StyledItems>
                            </Box>
                        ))}
                    </WarningContainer>
                }
            />
            <ActionsContainer>
                <Button
                    sx={{
                        position: 'static',
                        mr: 'auto',
                    }}
                    variant="outlined"
                    type="submit"
                    onClick={onBack}
                >
                    Back
                </Button>
                <PermissionButton
                    permission={CREATE_FEATURE}
                    projectId={project}
                    sx={{ position: 'static' }}
                    variant="contained"
                    type="submit"
                    onClick={onSubmit}
                    data-testid={IMPORT_CONFIGURATION_BUTTON}
                    disabled={
                        validationResult.errors.length > 0 ||
                        validationResult.permissions.length > 0 ||
                        !validJSON
                    }
                >
                    Import configuration
                </PermissionButton>
                <Button
                    sx={{ position: 'static', ml: 2 }}
                    variant="outlined"
                    type="submit"
                    onClick={onClose}
                >
                    Cancel import
                </Button>
            </ActionsContainer>
        </ImportLayoutContainer>
    );
};
