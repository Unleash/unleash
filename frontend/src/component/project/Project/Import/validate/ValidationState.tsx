import { ImportLayoutContainer } from '../ImportLayoutContainer';
import { Box, styled, Typography } from '@mui/material';
import { FC } from 'react';

const ImportInfoContainer = styled(Box)(({ theme }) => ({
    backgroundColor: theme.palette.secondaryContainer,
    borderRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(3),
}));

const Label = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.secondary,
}));
const Value = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.primary,
    fontWeight: theme.fontWeight.bold,
}));

export const ValidationStage: FC<{ environment: string; project: string }> = ({
    environment,
    project,
}) => {
    return (
        <ImportLayoutContainer>
            <ImportInfoContainer>
                <Typography sx={{ mb: 1.5 }}>
                    You are importing this configuration in:
                </Typography>
                <Box sx={{ display: 'flex', gap: 3 }}>
                    <span>
                        <Label>Environment: </Label>
                        <Value>{environment}</Value>
                    </span>
                    <span>
                        <Label>Project: </Label>
                        <Value>{project}</Value>
                    </span>
                </Box>
            </ImportInfoContainer>
        </ImportLayoutContainer>
    );
};
