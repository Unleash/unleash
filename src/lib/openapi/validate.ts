import Ajv, { ErrorObject } from 'ajv';
import { SchemaId, schemas } from './index';
import { omitKeys } from '../util/omit-keys';

interface ISchemaValidationErrors {
    schema: SchemaId;
    errors: ErrorObject[];
}

const ajv = new Ajv({
    schemas: Object.values(schemas).map((schema) =>
        omitKeys(schema, 'components'),
    ),

    // example was superseded by examples in openapi 3.1, but we're still on 3.0, so
    // let's add it back in!
    keywords: ['example'],
    formats: {
        'date-time': true,
        uri: true,
    },
});

export const validateSchema = (
    schema: SchemaId,
    data: unknown,
): ISchemaValidationErrors | undefined => {
    if (!ajv.validate(schema, data)) {
        return {
            schema,
            errors: ajv.errors ?? [],
        };
    }
};
