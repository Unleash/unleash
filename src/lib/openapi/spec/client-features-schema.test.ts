import { validateSchema } from '../validate';
import { ClientFeaturesSchema } from './client-features-schema';

test('clientFeaturesSchema no fields', () => {
    expect(
        validateSchema('#/components/schemas/clientFeaturesSchema', {}),
    ).toMatchSnapshot();
});

test('clientFeaturesSchema required fields', () => {
    const data: ClientFeaturesSchema = {
        version: 0,
        query: {},
        features: [
            {
                name: 'some-name',
                enabled: false,
                impressionData: false,
                variants: [
                    {
                        name: 'a',
                        weight: 1,
                        weightType: 'b',
                        stickiness: 'c',
                        payload: {
                            type: 'a',
                            value: 'b',
                        },
                        overrides: [
                            {
                                contextName: 'a',
                                values: ['b'],
                            },
                        ],
                    },
                ],
            },
        ],
    };

    expect(
        validateSchema('#/components/schemas/clientFeaturesSchema', data),
    ).toBeUndefined();
});

test('clientFeaturesSchema java-sdk expected response', () => {
    const json = `{
      "version": 2,
      "segments": [
        {
          "id": 1,
          "name": "some-name",
          "description": null,
          "constraints": [
            {
              "contextName": "some-name",
              "operator": "IN",
              "value": "name",
              "inverted": false,
              "caseInsensitive": true
            }
          ]
        }
      ],
      "features": [
        {
          "name": "Test.old",
          "description": "No variants here!",
          "enabled": true,
          "strategies": [
            {
              "name": "default"
            }
          ],
          "variants": null,
          "createdAt": "2019-01-24T10:38:10.370Z"
        },
        {
          "name": "Test.variants",
          "description": null,
          "enabled": true,
          "strategies": [
            {
              "name": "default",
              "segments": [
                1
              ]
            }
          ],
          "variants": [
            {
              "name": "variant1",
              "weight": 50
            },
            {
              "name": "variant2",
              "weight": 50
            }
          ],
          "createdAt": "2019-01-24T10:41:45.236Z"
        },
        {
          "name": "featureX",
          "enabled": true,
          "strategies": [
            {
              "name": "default"
            }
          ]
        },
        {
          "name": "featureY",
          "enabled": false,
          "strategies": [
            {
              "name": "baz",
              "parameters": {
                "foo": "bar"
              }
            }
          ]
        },
        {
          "name": "featureZ",
          "enabled": true,
          "strategies": [
            {
              "name": "default"
            },
            {
              "name": "hola",
              "parameters": {
                "name": "val"
              },
              "segments": [
                1
              ]
            }
          ]
        }
      ]
    }`;

    expect(
        validateSchema(
            '#/components/schemas/clientFeaturesSchema',
            JSON.parse(json),
        ),
    ).toBeUndefined();
});

test('clientFeaturesSchema unleash-proxy expected response', () => {
    const json = `{
      "version": 2,
      "segments": [
        {
          "id": 1,
          "name": "some-name",
          "description": null,
          "constraints": [
            {
              "contextName": "some-name",
              "operator": "IN",
              "value": "name",
              "inverted": false,
              "caseInsensitive": true
            }
          ]
        }
      ],
      "features": [
        {
          "name": "Test.old",
          "description": "No variants here!",
          "enabled": true,
          "strategies": [
            {
              "name": "default"
            }
          ],
          "variants": null,
          "createdAt": "2019-01-24T10:38:10.370Z"
        },
        {
          "name": "Test.variants",
          "description": null,
          "enabled": true,
          "strategies": [
            {
              "name": "default",
              "segments": [
                1
              ]
            }
          ],
          "variants": [
            {
              "name": "variant1",
              "weight": 50
            },
            {
              "name": "variant2",
              "weight": 50
            }
          ],
          "createdAt": "2019-01-24T10:41:45.236Z"
        },
        {
          "name": "featureX",
          "enabled": true,
          "strategies": [
            {
              "name": "default"
            }
          ]
        },
        {
          "name": "featureY",
          "enabled": false,
          "strategies": [
            {
              "name": "baz",
              "parameters": {
                "foo": "bar"
              }
            }
          ]
        },
        {
          "name": "featureZ",
          "enabled": true,
          "strategies": [
            {
              "name": "default"
            },
            {
              "name": "hola",
              "parameters": {
                "name": "val"
              },
              "segments": [
                1
              ]
            }
          ]
        }
      ]
    }`;

    expect(
        validateSchema(
            '#/components/schemas/clientFeaturesSchema',
            JSON.parse(json),
        ),
    ).toBeUndefined();
});

test('clientFeaturesSchema client specification test 15', () => {
    const json = `{
      "version": 2,
      "features": [
        {
          "name": "F9.globalSegmentOn",
          "description": "With global segment referencing constraint in on state",
          "enabled": true,
          "strategies": [
            {
              "name": "default",
              "parameters": {},
              "segments": [
                1
              ]
            }
          ]
        },
        {
          "name": "F9.globalSegmentOff",
          "description": "With global segment referencing constraint in off state",
          "enabled": true,
          "strategies": [
            {
              "name": "default",
              "parameters": {},
              "segments": [
                2
              ]
            }
          ]
        },
        {
          "name": "F9.globalSegmentAndConstraint",
          "description": "With global segment and constraint both on",
          "enabled": true,
          "strategies": [
            {
              "name": "default",
              "parameters": {},
              "constraints": [
                {
                  "contextName": "version",
                  "operator": "SEMVER_EQ",
                  "value": "1.2.2"
                }
              ],
              "segments": [
                1
              ]
            }
          ]
        },
        {
          "name": "F9.withExtraParams",
          "description": "With global segment that doesn't exist",
          "enabled": true,
          "project": "some-project",
          "strategies": [
            {
              "name": "default",
              "parameters": {},
              "constraints": [
                {
                  "contextName": "version",
                  "operator": "SEMVER_EQ",
                  "value": "1.2.2"
                }
              ],
              "segments": [
                3
              ]
            }
          ]
        },
        {
          "name": "F9.withSeveralConstraintsAndSegments",
          "description": "With several segments and constraints",
          "enabled": true,
          "strategies": [
            {
              "name": "default",
              "parameters": {},
              "constraints": [
                {
                  "contextName": "customNumber",
                  "operator": "NUM_LT",
                  "value": "10"
                },
                {
                  "contextName": "version",
                  "operator": "SEMVER_LT",
                  "value": "3.2.2"
                }
              ],
              "segments": [
                1,
                4,
                5
              ]
            }
          ]
        }
      ],
      "segments": [
        {
          "id": 1,
          "constraints": [
            {
              "contextName": "version",
              "operator": "SEMVER_EQ",
              "value": "1.2.2"
            }
          ]
        },
        {
          "id": 2,
          "constraints": [
            {
              "contextName": "version",
              "operator": "SEMVER_EQ",
              "value": "3.1.4"
            }
          ]
        },
        {
          "id": 3,
          "constraints": [
            {
              "contextName": "version",
              "operator": "SEMVER_EQ",
              "value": "3.1.4"
            }
          ]
        },
        {
          "id": 4,
          "constraints": [
            {
              "contextName": "customName",
              "operator": "STR_CONTAINS",
              "values": [
                "Pi"
              ]
            }
          ]
        },
        {
          "id": 5,
          "constraints": [
            {
              "contextName": "slicesLeft",
              "operator": "NUM_LTE",
              "value": "4"
            }
          ]
        }
      ]
    }`;

    expect(
        validateSchema(
            '#/components/schemas/clientFeaturesSchema',
            JSON.parse(json),
        ),
    ).toBeUndefined();
});
