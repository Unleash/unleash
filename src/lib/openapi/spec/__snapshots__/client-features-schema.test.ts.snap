// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clientFeaturesSchema no fields 1`] = `
{
  "errors": [
    {
      "instancePath": "",
      "keyword": "required",
      "message": "must have required property 'version'",
      "params": {
        "missingProperty": "version",
      },
      "schemaPath": "#/required",
    },
  ],
  "schema": "#/components/schemas/clientFeaturesSchema",
}
`;
